templates/commands/render_action.go

```diff
@@ -40,8 +40,16 @@ type walkAndModifyVisitor func([]byte) ([]byte, error)
 // Recursively traverses the directory or file scratchDir/relPath, calling the
 // given visitor for each file. If the visitor returns modified file contents
 // for a given file, that file will be overwritten with the new contents.
-func walkAndModify(ctx context.Context, pos *model.ConfigPos, rfs renderFS, scratchDir, relPath string, v walkAndModifyVisitor) error {
+// The seen map ensures that the same file isn't visited multiple times in subsequent
+// calls. It must be non-nil.
+func walkAndModify(ctx context.Context, pos *model.ConfigPos, rfs renderFS, scratchDir, relPath string, v walkAndModifyVisitor, seen map[string]struct{}) error {
 	logger := logging.FromContext(ctx).Named("walkAndModify")
+	if seen == nil {
+		// Intentionally forcing caller to initialize map to discourage them from just calling with nil and not using
+		// the map for future calls.
+		return fmt.Errorf("to avoid handling the same file multiple times, the seen map provided to " +
+			"walkAndModify must be non-nil")
+	}
```

Comment:
Why do they need to pass in a map at all? Shouldn't the `seen` map be local to the walkAndModify function? Or is it because this is called for each action?

Reply:
Yeah, its called for each function. Within a particular call to walkAndModify there is no danger of multiple passes (unless something evil is done with symlinks, which this won't solve either).

The danger comes with subsequent calls, so some information needs to be preserved between them.

Reply:
An alternate solution would be to change walkAndModify to accept multiple starting paths.

Reply:
I think passing a list of paths to walkAndModify would be clean, and I don't see any reason why it wouldn't work.

Reply:
Changed